spark-cassandra-connector/src/test/scala/org/apache/spark/sql/cassandra/PredicatePushDownSpec.scala

```diff
@@ -0,0 +1,241 @@
+    ppd.predicatesToPreserve shouldBe Seq(f2)
+  }
+
+  it should "push down clustering column predicates when the last clustering column is restricted by IN" in {
+    val f1 = EqFilter("c1")
+    val f2 = EqFilter("c2")
+    val f3 = InFilter("c3")
+    val ppd = new PredicatePushDown(Seq[Filter](f1, f2, f3), table)
+    ppd.predicatesToPushDown should contain only(f1, f2, f3)
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "stop pushing down clustering column predicates on the first range predicate" in {
+    val f1 = EqFilter("c1")
+    val f2 = RangeFilter("c2")
+    val f3 = EqFilter("c3")
+    val ppd = new PredicatePushDown(Seq[Filter](f1, f2, f3), table)
+    ppd.predicatesToPushDown should contain only(f1, f2)
+    ppd.predicatesToPreserve shouldBe Seq(f3)
+  }
+
+  it should "not push down IN restriction on non-last column" in {
+    val f1 = EqFilter("c1")
+    val f2 = InFilter("c2")
+    val f3 = EqFilter("c3")
+    val ppd = new PredicatePushDown(Seq[Filter](f1, f2, f3), table)
+    ppd.predicatesToPushDown should contain only f1
+    ppd.predicatesToPreserve should contain only (f2, f3)
+  }
+
+  it should "not push down any clustering column predicates, if the first clustering column is missing" in {
+    val f1 = EqFilter("c2")
+    val ppd = new PredicatePushDown(Seq[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve shouldBe Seq(f1)
+  }
+
+  it should "push down equality predicates on regular indexed columns" in {
+    val f1 = EqFilter("i1")
+    val ppd = new PredicatePushDown(Seq[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe Seq(f1)
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "not push down range predicates on regular indexed columns" in {
+    val f1 = RangeFilter("i1")
+    val ppd = new PredicatePushDown(Seq[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve shouldBe Seq(f1)
+  }
+
+  it should "not push down IN predicates on regular indexed columns" in {
+    val f1 = InFilter("i1")
+    val ppd = new PredicatePushDown(Seq[Filter](f1), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve shouldBe Seq(f1)
+  }
+
+  it should "push down predicates on regular non-indexed and indexed columns" in {
+    val f1 = EqFilter("r1")
+    val f2 = EqFilter("r2")
+    val f3 = EqFilter("i1")
+    val ppd = new PredicatePushDown(Seq[Filter](f1, f2, f3), table)
+    ppd.predicatesToPushDown should contain allOf(f1, f2, f3)
+    ppd.predicatesToPreserve shouldBe empty
+  }
+
+  it should "not push down predicates on regular non-indexed columns if indexed ones are not included" in {
+    val f1 = EqFilter("r1")
+    val f2 = EqFilter("r2")
+    val ppd = new PredicatePushDown(Seq[Filter](f1, f2), table)
+    ppd.predicatesToPushDown shouldBe empty
+    ppd.predicatesToPreserve should contain allOf(f1, f2)
+  }
+
+  it should "not rely on given predicates order (1)" in {
```

Comment:
perhaps you can give more details about the test scenario - instead of adding 1, 2, 3 suffix